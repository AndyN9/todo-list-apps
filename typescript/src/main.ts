import { v4 as uuidV4 } from 'uuid';

import './style.css'

const list = document.querySelector<HTMLUListElement>('#todo-list');
const form = document.querySelector<HTMLFormElement>('#new-todo-form');
const input = document.querySelector<HTMLInputElement>('#new-todo-task');

type Task = {
  id: string;
  title: string;
  completed: boolean;
}

let tasks: Task[] = loadTasks();
renderList();

form?.addEventListener('submit', event => {
  event.preventDefault();

  if (input?.value === "" || input?.value == null) {

    return;
  }

  const newTask: Task = {
    id: uuidV4(),
    title: input.value,
    completed: false,
  }

  tasks.push(newTask);
  addListItem(newTask);
  input.value = "";
  saveTasks();
});

function renderList(): void {
  if (list === null) {

    return;
  }

  list.innerHTML = "";
  tasks.forEach(addListItem);
}

function addListItem(task: Task): void {
  const listItem = document.createElement('li');
  const anchor = document.createElement('a');
  const label = document.createElement('label');
  const checkbox = document.createElement('input');

  anchor.innerText = 'Remove';
  anchor.addEventListener('click', () => {
    tasks = tasks.filter((item) => {

      return item.id !== task.id;
    });
    saveTasks();
    renderList();
  });

  checkbox.type = 'checkbox';
  checkbox.checked = task.completed;
  checkbox.addEventListener('change', () => {
    task.completed = checkbox.checked;
    saveTasks();
  });

  label.append(checkbox, task.title);
  listItem.append(label, anchor);
  list?.append(listItem);
}

function loadTasks(): Task[] {
  const taskJson = localStorage.getItem('TASKS');
  if (taskJson === null) {

    return [];
  }

  return JSON.parse(taskJson);
}

function saveTasks(): void {
  localStorage.setItem('TASKS', JSON.stringify(tasks));
}
