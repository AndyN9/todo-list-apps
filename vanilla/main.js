import { v4 as uuidV4 } from 'uuid';

import './style.css'

const list = document.querySelector('#todo-list');
const form = document.querySelector('#new-todo-form');
const input = document.querySelector('#new-todo-task');
export const LOCAL_STORAGE_KEY = 'TASKS';

let tasks = loadTasks();
renderList();

form?.addEventListener('submit', event => {
  event.preventDefault();

  if (input?.value === "" || input?.value == null) {

    return;
  }

  const newTask = {
    id: uuidV4(),
    title: input.value,
    completed: false,
  }

  tasks.push(newTask);
  addListItem(newTask);
  input.value = "";
  saveTasks(tasks);
});

function renderList() {
  if (list === null) {

    return;
  }

  list.innerHTML = "";
  tasks.forEach(addListItem);
}

function addListItem(task) {
  const listItem = document.createElement('li');
  const anchor = document.createElement('a');
  const label = document.createElement('label');
  const checkbox = document.createElement('input');

  anchor.innerText = 'Remove';
  anchor.addEventListener('click', () => {
    tasks = tasks.filter((item) => {

      return item.id !== task.id;
    });
    saveTasks(tasks);
    renderList();
  });

  checkbox.type = 'checkbox';
  checkbox.checked = task.completed;
  checkbox.addEventListener('change', () => {
    task.completed = checkbox.checked;
    saveTasks(tasks);
  });

  label.append(checkbox, task.title);
  listItem.append(label, anchor);
  list?.append(listItem);
}

export function loadTasks() {
  const taskJson = localStorage.getItem(LOCAL_STORAGE_KEY);
  if (taskJson === null || typeof taskJson === 'undefined') {

    return [];
  }

  return JSON.parse(taskJson);
}

export function saveTasks(tasks) {
  localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(tasks));
}
