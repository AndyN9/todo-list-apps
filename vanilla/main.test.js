import fs from 'fs';
import path from 'path';
import { v4 as uuidV4 } from 'uuid';
import { vi, describe, it, expect, beforeEach } from 'vitest';

import { LOCAL_STORAGE_KEY, loadTasks, saveTasks } from './main';

describe('app elements', () => {
  const html = fs.readFileSync(path.resolve(__dirname, './index.html'));

  beforeEach(() => {
    document.documentElement.innerHTML = html.toString();
  });

  it('should have correct elements', () => {
    const appElement = document.querySelector('#app');
    expect(appElement).toMatchSnapshot();
  });

  it('should match page title & header with app name', () => {
    const appName = 'Vanilla Todo List';
    const titleElement = document.querySelector('title');
    const headerElement = document.querySelector('h1');

    // innerText returns 'undefined'?
    expect(titleElement.innerHTML).toBe(appName);
    expect(headerElement.innerHTML).toBe(appName);
  });
});

describe('localStorage functions', () => {
  const initialTasks = [
    {
      id: uuidV4(),
      title: 'Create data',
      completed: false,
    },
    {
      id: uuidV4(),
      title: 'Read data',
      completed: false,
    },
    {
      id: uuidV4(),
      title: 'Update data',
      completed: false,
    },
    {
      id: uuidV4(),
      title: 'Delete data',
      completed: false,
    },
  ];
  const localStorageMock = (function () {
    let store = {};

    return {
      getItem(key) {
        return store[key];
      },

      setItem(key, value) {
        store[key] = value;
      },

      clear() {
        store = {};
      },

      removeItem(key) {
        delete store[key];
      },

      getAll() {
        return store;
      },
    };
  })();
  vi.stubGlobal('localStorage', localStorageMock);

  beforeEach(() => {
    localStorage.clear();
  });

  it('should load empty initial tasks', () => {
    const emptyLoadedInitialTasks = loadTasks();
    expect(emptyLoadedInitialTasks).toStrictEqual([]);
  });

  it('should load initial tasks', () => {
    // initial browser load
    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(initialTasks));

    const tasks = loadTasks();
    expect(tasks).toStrictEqual(initialTasks);
  });

  it('should save tasks', () => {
    saveTasks(initialTasks);

    const tasks = loadTasks();
    expect(tasks).toStrictEqual(initialTasks);
  });
});
