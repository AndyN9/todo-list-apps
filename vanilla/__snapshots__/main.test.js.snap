// Vitest Snapshot v1

exports[`app elements > should render correct elements 1`] = `
<div
  id="app"
>
  
    
  <h1>
    Vanilla Todo List
  </h1>
  
    
  <ul
    id="todo-list"
  />
  
    
  <form
    id="new-todo-form"
  >
    
      
    <input
      aria-label="Enter in new todo task"
      id="new-todo-task"
      type="text"
    />
    
      
    <button
      type="submit"
    >
      Add
    </button>
    
    
  </form>
  
  
</div>
`;
